### PR TITLE
Enable prescribed ice for UFS

### PR DESCRIPTION
**Pull request by NickSzapiro-NOAA**
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [x] Short (1 sentence) summary of your PR: 
    Enable prescribed ice CICE for UFS in nuopc/cmeps cap
- [x] Developer(s): 
    Nicholas Szapiro & Denise Worthen
- [ ] Suggest PR reviewers from list in the column to the right.
- [x] Please copy the PR test results link or provide a summary of testing completed below.
   Regression test atm_ds2s_docn_pcice added to ufs-weather-model (https://github.com/ufs-community/ufs-weather-model/pull/2186) 
- How much do the PR code changes differ from the unmodified code? 
    - [x] bit for bit (when not in prescribed ice)
    - [ ] different at roundoff level
    - [x] more substantial (when in prescribed ice #ifndef cesmcoupled) 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [x] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [x] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [x] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [x] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [x] No 
- [x] Please document the changes in detail, including _why_ the changes are made.  This will become part of the PR commit log.
Prescribed ice CICE was enclosed by #ifdef CESMCOUPLED in CMEPS driver. This allows use of prescribed ice CICE in ufs-weather-model with CDEPS calls to input (sea ice concentration) from file. Starting as cherry-picked as in UFS but dependency on CDEPS likely needs to be optional at compile time. This is needed for https://github.com/NOAA-EMC/CICE/issues/84

### CODE REVIEW

**Comment by NickSzapiro-NOAA** [comment]
The dependency on CDEPS is not a problem in UFS, but I'm not sure the best way to handle it here. Maybe a different #ifdef is needed in cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90 ?

File: cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90
Comment (dabail10) [inline cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90:18]: These come from the CIME shr stuff I believe. This is how we do the reading from stream files.

**Review by dabail10 — COMMENTED** [review]
This looks fine as far as I can tell. I will run with it on Monday to make sure.

**Comment by NickSzapiro-NOAA** [comment]
Thanks Dave! So most of the code in cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90 was enclosed in an #ifndef CESMCOUPLED else. There are the functions now "uncovered" to use the CDEPS inline functionality:
 https://github.com/ufs-community/ufs-weather-model/pull/2186/files#diff-db20ac41c986cab4cffeeaf178a6a4abf1c092a137af4cbe274e19deb7326954

**Comment by anton-seaice** [comment]
If I am reading this correctly, there were already some depedencies on CDEPS (i.e. other `use dshr_` statements) in this file and no-one has raised it as an issue? So I think we can just leave it be until someone wants to build without CDEPS? I have a vague recollection of someone saying there are other unexpected dependencies between CMEPS & CDEPS in places, so I don't think this would break anything particularly.

**Comment by dabail10** [comment]
There was only a dependency on CDEPS when CESMCOUPLED was defined. I guess my concern is that we have code duplication here. Will we need two copies of the CDEPS library? What happens when they get out of sync?

File: cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90
Comment (NickSzapiro-NOAA) [inline cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90:213]: UFS needed this call ice_prescribed_run at the end of ice_prescribed_init. I don't know if it's a bug for CESM not to. Presumably, the CICE state should be initialized in init, but it may depend on the components' run sequence. We left the #ifndef CESMCOUPLED so as not to impact CESM @dabail10

**Comment by NickSzapiro-NOAA** [comment]
The ice prescribed code is only useful if CICE is linked with CDEPS. I didn't think including CDEPS with CICE was an option. I'm happy to block it all off if not...maybe #ifdef CICE_HAS_CDEPS (?)

**Comment by dabail10** [comment]
I guess I am confused. I understand that you want to use the prescribed ice and hence CDEPS. What I think you are saying is that you would have a separate set of modules / subroutines from CDEPS that would come with CICE. Kind of like what is done for the shr_orb routines. When we use CICE in CESM, then CDEPS is part of the CESM checkout. One thing you might be able to do is have CDEPS as a submodule / external for CICE. This is only checked out when using UFS. Perhaps I am missing something still?

**Comment by NickSzapiro-NOAA** [comment]
Sorry for being confusing.

UFS already has CDEPS as a subcomponent. This code builds and runs fine in UFS (with appropriate linking). It shouldn't change anything for CESM either.

I was worried if anyone else uses this driver and does not have CDEPS. If that is not a concern, I agree with @anton-seaice that this code is fine as is until further notice. No stubs (or CDEPS submodule) are needed here then

File: cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90
Comment (anton-seaice) [inline cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90:13]: I think we should leave this section, and add a UFSCOUPLED or a NO_CDEPS FLAG .

**Comment by anton-seaice** [comment]
> There was only a dependency on CDEPS when CESMCOUPLED was defined. I guess my concern is that we have code duplication here. 

Sorry - you are right. I didn't realise most of the ice_prescribed_mod file was within the CESM coupled ifdef. 

I think we shouldn't do something that might require others to add a compiler flag if there existing setup works ... ? I added a comment to this effect - https://github.com/CICE-Consortium/CICE/pull/964/files#r1708080117

**Comment by dabail10** [comment]
I like this strategy of adding a NOCDEPS flag or something similar. However, how many groups use the nuopc/cmeps driver? If it is only the three of us and we all use CDEPS, then I would advocate for removing the CESMCOUPLED flag here.

**Comment by DeniseWorthen** [comment]
> I like this strategy of adding a NOCDEPS flag or something similar. However, how many groups use the nuopc/cmeps driver? If it is only the three of us and we all use CDEPS, then I would advocate for removing the CESMCOUPLED flag here.

I don't think any ifdef is required in the ``ice_prescribed_mod`` at this point. If someone were to attempt to use this mode w/o CDEPS, they would probably either a) fail compile or b) fail at run time when the ``ice_prescribed_init ``routine tries to call the ``dshr_pio_init`` or ``shr_strdata_init_from_inline``.

**Comment by dabail10** [comment]
We should probably add an informative message here to let people know that CDEPS is required.

**Comment by NickSzapiro-NOAA** [comment]
> We should probably add an informative message here to let people know that CDEPS is required.

I don't see any docs for drivers/nuopc/cmeps . May I ask where I should inform of this dependency on CDEPS?

**Comment by dabail10** [comment]
There is no documentation for the driver per se. I think we just need a graceful abort that checks the first call to CDEPS and prints a message saying that CDEPS is required for this module or something like that.

**Comment by NickSzapiro-NOAA** [comment]
Using this code as is without any "substitute" ifdefs, the nuopc/cmeps driver code should fail to compile without CDEPS. This isn't a practical problem as CDEPS comes with CMEPS in coupled models we know of. @anton-seaice, do y'all share this driver too?

It is useful to enable this prescribed ice code more generally beyond CESMCOUPLED. 

Maybe the position here is: Why "clutter" the code with ifdefs, stub routines, etc. unless needed? Especially if just to abort. This is of course a Consortium decision and open to other solutions.

**Comment by dabail10** [comment]
Excellent point. So, I agree with removing the ifdefs and it is just buyer beware for those using the nuopc/cmeps driver.  Maybe something in the comments about the dependency on CDEPS? That way if it fails to compile, someone could look in ice_prescribed_mod.F90 and see this note.

**Comment by apcraig** [comment]
My feeling is that if the nuopc/cmeps implementation is always used with cdeps as far as we know, then get rid of the CPPs.  Especially if they are not tested.  If someone comes along in the future that uses nuopc/cmeps and doesn't have cdeps (and doesn't need prescribed ice mode), they can add the CPPs and test it at that time.

File: cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90
Comment (NickSzapiro-NOAA) [reply cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90:13]: It seems cleaner to admit the dependency on CDEPS. Is this ok @anton-seaice?

Comment (anton-seaice) [reply cicecore/drivers/nuopc/cmeps/ice_prescribed_mod.F90:13]: Consensus seems to be to go with the PR as currently written

**Review by anton-seaice — APPROVED** [review]
I checked this builds and starts, we don't use prescribed_ice so I didn't test that.

**Comment by apcraig** [comment]
Is this ready to merge?  @dabail10, can you review/confirm when you have a chance.  It's fine by me but have no stake in and have not tested these changes.  Thanks.

**Comment by dabail10** [comment]
I have not had time to test this. I think it is fine.

**Comment by apcraig** [comment]
@dabail10.  I'll wait for you to approve.  I'm not in a rush.  Will leave it up to you whether to test or not before merging.

**Comment by dabail10** [comment]
Just tested this in CESM and it passed a simple debug restart test.

**Review by dabail10 — APPROVED** [review]

**Review by apcraig — APPROVED** [review]

**Pull request merged**
